src/actions/tests/pending-breakpoints.js

```diff
@@ -0,0 +1,261 @@
+  return {
+    location: {
+      sourceUrl: `http://localhost:8000/examples/${filename}`,
+      sourceId: filename,
+      line: 5
+    },
+    condition: null,
+    disabled: false
+  };
+}
+
+function generatePendingBreakpoint(breakpoint) {
+  const {
+    location: { sourceUrl, line, column },
+    condition,
+    disabled
+  } = breakpoint;
+
+  return {
+    location: { sourceUrl, line, column },
+    condition,
+    disabled
+  };
+}
+
+function slideMockBp(bp) {
+  const slidBp = Object.assign({}, bp);
+  slidBp.location.line = bp.location.line + 2;
+  return slidBp;
+}
+
+const slidingMockThreadClient = {
+  setBreakpoint: (location, condition) => {
+    return new Promise((resolve, reject) => {
+      const actualLocation = Object.assign({}, location, {
+        line: location.line + 2
+      });
+      resolve({ id: makeLocationId(location), actualLocation, condition });
+    });
+  }
+};
+
+const simpleMockThreadClient = {
+  setBreakpoint: (location, condition) => {
+    return new Promise((resolve, reject) => {
+      resolve({ id: "hi", actualLocation: location });
+    });
+  },
+
+  removeBreakpoint: id => {
+    return new Promise((resolve, reject) => {
+      resolve({ status: "done" });
+    });
+  },
+
+  setBreakpointCondition: (id, location, condition, noSliding) => {
+    return new Promise((resolve, reject) => {
+      resolve({ sourceId: "a", line: 5 });
+    });
+  }
+};
+
+describe("pending breakpoints", () => {
+  it("when the user adds a breakpoint, a corresponding pending breakpoint should be added", async () => {
+    const { dispatch, getState } = createStore(simpleMockThreadClient);
+    const bp = generateBreakpoint("foo");
+    const id = makePendingLocationId(bp.location);
+
+    await dispatch(actions.addBreakpoint(bp.location));
+    const pendingBps = selectors.getPendingBreakpoints(getState());
+    expect(pendingBps.size).to.be(2);
+    expect(pendingBps.get(id)).to.eql(generatePendingBreakpoint(bp));
+  });
+
+  it("when the user adds a sliding breakpoint, a corresponding pending breakpoint should be added", async () => {
+    const { dispatch, getState } = createStore(slidingMockThreadClient);
+    const bp = generateBreakpoint("foo");
+    await dispatch(actions.addBreakpoint(bp.location));
+    const pendingBps = selectors.getPendingBreakpoints(getState());
+
+    const slidBp = slideMockBp(bp);
+    const newId = makePendingLocationId(slidBp.location);
+
+    const bps = selectors.getPendingBreakpoints(getState());
+    const newBp = pendingBps.get(newId);
+    expect(newBp).to.eql(generatePendingBreakpoint(slidBp));
+  });
+
+  describe("when two or more breakpoints are added", () => {
+    let breakpoint1;
+    let breakpoint2;
+    let breakpointId1;
+    let breakpointId2;
+
+    beforeEach(() => {
+      breakpoint1 = generateBreakpoint("foo");
+      breakpoint2 = generateBreakpoint("foo2");
+      breakpointId1 = makePendingLocationId(breakpoint1.location);
```

Comment:
nitpick: now that I understand the relationship between `locationId` and `id` in the breakpoints, maybe this should be rewritten as `breakpointLocationId1` and `breakpointLocationId2`